subsys/bluetooth/host/scan.h

```diff
@@ -5,12 +5,81 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#ifndef SUBSYS_BLUETOOTH_HOST_SCAN_H_
+#define SUBSYS_BLUETOOTH_HOST_SCAN_H_
+
+#include <stdint.h>
+
+#include <zephyr/sys/atomic.h>
+#include <zephyr/bluetooth/bluetooth.h>
+
+/**
+ * Reasons why a scanner can be running.
+ * Used as input to @ref bt_le_scan_update_and_reconfigure.
+ */
+enum bt_le_scan_enabled_reason {
+	/** The application explicitly instructed the stack to scan for advertisers
+	 * using the API @ref bt_le_scan_start().
+	 * May not be used explicitly.
+	 */
+	BT_LE_SCAN_ENABLED_REASON_EXPLICIT_SCAN,
+
+	/**
+	 * Syncing to a periodic advertiser.
+	 */
+	BT_LE_SCAN_ENABLED_REASON_SYNC_SYNCING,
+
+	/**
+	 * Scanning to find devices to connect to.
+	 */
+	BT_LE_SCAN_ENABLED_REASON_SCAN_BEFORE_INITIATE,
+
+	/**
+	 * Special state for a NOP for @ref bt_le_scan_update_and_reconfigure to not add/remove any
+	 * reason why the scanner should be running.
+	 */
+	BT_LE_SCAN_ENABLED_REASON_NONE,
+	BT_LE_SCAN_ENABLED_REASON_NUM_FLAGS,
+};
+
 void bt_scan_reset(void);
 
 bool bt_id_scan_random_addr_check(void);
+bool bt_le_scan_active_scanner_running(void);
 
 int bt_le_scan_set_enable(uint8_t enable);
 
 struct bt_le_per_adv_sync *bt_hci_get_per_adv_sync(uint16_t handle);
 
 void bt_periodic_sync_disable(void);
+
+/**
+ * Start / stop / update the scanner.
+ *
+ * This API updates the reasons why the scanner is running.
+ * Depending on all the reasons why the scanner is enabled, scan parameters are selected
+ * and the scanner is started or stopped, if needed.
+ * This API may update the scan parameters, for example if the scanner is already running
+ * when another reason that demands higher duty-cycle is being added.
+ *
+ * This API also allows removing the reason why the scanner is running.
+ * To do that, the enable_reason has to be set to the same value that was used to enable the
+ * scanner. This is necessary because different parts of the host may start the scanner and
+ * these parts don't need to know if other parts of the host or the application already started
+ * the scanner. This means that every SW module that informs the scanner that it should run,
+ * needs to eventually stop it with the same reason.
+ * If all reasons for the scanner to run are removed, this API will stop the scanner.
+ *
+ * @param enable_reason reason why the scanner should be running
+ * @param enable enable or disable the reason in previous argument. Use @ref BT_HCI_LE_SCAN_ENABLE
+ *               or @ref BT_HCI_LE_SCAN_DISABLE.
+ *
+ * @return 0 in case of success, or a negative error code on failure.
+ */
+int bt_le_scan_update_and_reconfigure(enum bt_le_scan_enabled_reason enable_reason, uint8_t enable);
```

Comment:
I had a comment about 
```c
				bt_le_scan_update_and_reconfigure(
					BT_LE_SCAN_ENABLED_REASON_SCAN_BEFORE_INITIATE,
					BT_HCI_LE_SCAN_DISABLE);
```
and why we set an enable reason when we disable, but re-reading this, it sounds like the second parameter, `BT_HCI_LE_SCAN_DISABLE` is not whether we are stating/stopping scanning, but whether the reason, `BT_LE_SCAN_ENABLED_REASON_SCAN_BEFORE_INITIATE` is set or unset. Is that correct?

In which case then I think the naming can be improved. We could use the naming convention used by the flags:
```suggestion
int bt_le_scan_set_reason(enum bt_le_scan_enabled_reason enable_reason);
int bt_le_scan_unset_reason(enum bt_le_scan_enabled_reason enable_reason);
/* or */
int bt_le_scan_clear_reason(enum bt_le_scan_enabled_reason enable_reason);
```

I might also be missing some overall design idea of using the term `reason` when this looks more like a state machine, and why the term `state` was not used. 

Can you elaborate on that?

Reply:
I think the problem for the naming is that it is kind of a state machine but also it's not because there are many possible states.
Multiple "reasons" can be enabled at the same time, so they are not exactly states. The states in the state-machine would be all possible combinations of "reasons".

I like your idea of splitting the API. However, I would like to update the naming to tell something about the side-effect. The side-effect is that the scanner might be started, stopped or updated.

What do you think about 
```suggestion
int bt_le_scan_set_reason_and_update(enum bt_le_scan_enabled_reason enable_reason);
int bt_le_scan_unset_reason_and_update(enum bt_le_scan_enabled_reason enable_reason);
```
?

Reply:
Naming is hard :D 

Not a fan of "reason" either, as you say there can be multiple. Semantically (at least in my head), it doesn't make sense to have multiple "reasons" enabled at any one point. 

An obvious alternative is to simply use a more generic term like "flag" (or "scan_flag") which is more common. 

There may be a better name that so I'm more than open for ideas :)

But other than that, the PR looks good now and it's a welcome change